app/controllers/application_controller.rb

```diff
@@ -5,6 +5,148 @@ class ApplicationController < Sinatra::Base
+  end
+
+### SIGNUP ###
+  get '/signup' do 
+    if Helpers.is_logged_in?(session)
+      redirect "/tweets"
+    else
+      erb :'/users/create_user'
+    end
+  end
+
+  post '/signup' do 
+    @user = User.new(username: params["username"], email: params["email"], password: params["password"])
+    if !@user.username.empty? && !@user.email.empty? && @user.password != nil
+      @user.save
+      session[:user_session_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+    end
+  end
+
+### LOGIN ###
+  get '/login' do
+    if Helpers.is_logged_in?(session)
+      redirect "/tweets"
+    else
+      erb :'/users/login'
+    end
+  end
+
+  post '/login' do 
+    @user = User.find_by(:username => params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:user_session_id] = @user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+### LOGOUT ###
+  get '/logout' do
+    if Helpers.is_logged_in?(session)
+      session.clear
+      redirect "/login"
+    else
+      redirect "/"
+    end
+  end
+
+  ### SHOW ALL TWEETS OF SPECIFIC USER ###
+  get '/users/:slug' do 
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show_tweets'
+  end
+
+
+### CREATE NEW TWEET ###
+  get '/tweets/new' do
+    if Helpers.is_logged_in?(session)
+      erb :'/tweets/create_tweet' 
+    else
+      redirect "/login"
+    end
+  end
+
+  post '/tweets/new' do 
+    @tweet = Tweet.new(:content => params["content"], :user_id => params["user_id"])
+    if !@tweet.content.empty?
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/new"
+    end
+  end
+
+### SHOW ALL TWEETS ###
+  get '/tweets' do 
+    if Helpers.is_logged_in?(session)
+      erb :'tweets/tweets'
+    else
+      redirect "/login"
+    end
+  end
+
+  ### SHOW SPECIFIC TWEETS ###
+    get '/tweets/:id' do 
+      if Helpers.is_logged_in?(session)
+        @tweet = Tweet.find(params[:id])
+        erb :'tweets/show_tweet'
+      else
+        redirect "/login"
+      end
+    end
+
+### EDIT SPECIFIC TWEETS ###
+  get '/tweets/:id/edit' do 
+    if Helpers.is_logged_in?(session)
+      @inquiring_user = Helpers.current_user(session)
+      @tweet = Tweet.find(params[:id])
+        if @inquiring_user.id == @tweet.user_id
+          erb :'/tweets/edit_tweet'
+        else
+          redirect "/tweets"
+        end
+    else
+      redirect "/login"
+    end
+  end
+
+  patch '/tweets/:id' do 
+    if params["content"].empty?
+      redirect "/tweets/#{params[:id]}/edit"
+    else
+      @tweet = Tweet.find(params[:id])
+      @tweet.content = params["content"]
+      @tweet.save
+      redirect "/tweets"
+    end
+  end
+
+  ### DELETE SPECIFIC TWEET
+  delete '/tweets/:id/delete' do 
+    if Helpers.is_logged_in?(session)
+      @user = Helpers.current_user(session)
+      @tweet = Tweet.find(params[:id])
+      if @user.id == @tweet.id
+        @tweet.destroy
```

Comment:
Nice!  There it is again, I didn't realize I could call something like find_by on current_user.tweets.

It's geeky AF to say, but ActiveRecord is pretty awesome sometimes.


> On Apr 9, 2016, at 7:49 PM, Randall Reed <notifications@github.com> wrote:
> 
> In app/controllers/application_controller.rb <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/79#discussion_r59125802>:
> 
> > +      redirect "/tweets/#{params[:id]}/edit"
> > +    else
> > +      @tweet = Tweet.find(params[:id])
> > +      @tweet.content = params["content"]
> > +      @tweet.save
> > +      redirect "/tweets"
> > +    end
> > +  end
> > +
> > +  ### DELETE SPECIFIC TWEET
> > +  delete '/tweets/:id/delete' do 
> > +    if Helpers.is_logged_in?(session)
> > +      @user = Helpers.current_user(session)
> > +      @tweet = Tweet.find(params[:id])
> > +      if @user.id == @tweet.id
> > +        @tweet.destroy
> Prefer to find tweets that are already only associated to the current_user than to manually check the user_id foreign key of the tweet against the current_user.id
> 
> tweet = current_user.tweets.find_by(id: params[:id])
> if tweet && tweet.destroy
>   redirect "/tweets"
> else
>   redirect "/tweets/#{tweet.id}"
> end
> We're using the association collection method provided by has_many <http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many> to query the user's tweets.
> 
> Note the usage of find_by <http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by> vs find <http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find>. Had we used find here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. find_by will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/79/files/9dfc20de18dce246a6925028c350c9316e9e3937#r59125802>